core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionRepository.java

```diff
@@ -23,53 +30,113 @@
 
 /**
  * Builtin Function Repository.
+ * Repository registers catalog specific functions under catalog specific namespace and
+ * universal functions under default namespace. Catalog Specific Namespace carries their own
+ * namespace.
+ *
  */
 @RequiredArgsConstructor
 public class BuiltinFunctionRepository {
-  private final Map<FunctionName, FunctionResolver> functionResolverMap;
+
+  public static final String DEFAULT_NAMESPACE = "default";
+
+  private final Map<String, Map<FunctionName, FunctionResolver>> namespaceFunctionResolverMap;
+
 
   /**
-   * Register {@link DefaultFunctionResolver} to the Builtin Function Repository.
+   * Register {@link DefaultFunctionResolver} to the Builtin Function Repository
+   * under default namespace.
    *
    * @param resolver {@link DefaultFunctionResolver} to be registered
    */
   public void register(FunctionResolver resolver) {
-    functionResolverMap.put(resolver.getFunctionName(), resolver);
+    register(DEFAULT_NAMESPACE, resolver);
   }
 
   /**
-   * Compile FunctionExpression.
+   * Register {@link DefaultFunctionResolver} to the Builtin Function Repository with
+   * specified namespace.
+   *
+   * @param resolver {@link DefaultFunctionResolver} to be registered
+   */
+  public void register(String namespace, FunctionResolver resolver) {
+    Map<FunctionName, FunctionResolver> functionResolverMap;
+    if (!namespaceFunctionResolverMap.containsKey(namespace)) {
+      functionResolverMap = new HashMap<>();
+      namespaceFunctionResolverMap.put(namespace, functionResolverMap);
+    }
+    namespaceFunctionResolverMap.get(namespace).put(resolver.getFunctionName(), resolver);
+  }
+
+
+  /**
+   * Compile FunctionExpression under default namespace.
+   *
    */
   public FunctionImplementation compile(FunctionName functionName, List<Expression> expressions) {
-    FunctionBuilder resolvedFunctionBuilder = resolve(new FunctionSignature(functionName,
-        expressions.stream().map(expression -> expression.type()).collect(Collectors.toList())));
+    return compile(DEFAULT_NAMESPACE, functionName, expressions);
+  }
+
+
+  /**
+   * Compile FunctionExpression within given namespace.
+   * Checks for default namespace first and then tries to compile from given namespace.
+   */
+  public FunctionImplementation compile(String namespace, FunctionName functionName,
```

Comment:
1. namespace is catalog name?
2. if yes, the Opensearch storage engine should implement StorageEngine.getFunctions interface also.  and reuse the FunctionResolver logic.
3. if yes, we should add catalog name to FunctionName.

Reply:
https://github.com/opensearch-project/sql/issues/930